File: frontend/solutions/p9/aiSolution.py
class Solution:
    def numberOfSets(self, n: int, k: int) -> int:
        MOD = 10 ** 9 + 7
        # dp[i] = ways to place the current number of segments using points 0..i
        dp = [1] * n

        for _ in range(k):
            new_dp = [0] * n
            prefix = 0
            for i in range(n):
                new_dp[i] = ((new_dp[i - 1] if i > 0 else 0) + prefix) % MOD
                prefix = (prefix + dp[i]) % MOD
            dp = new_dp

        return dp[-1] % MOD

File: frontend/solutions/p9/aiSolution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var numberOfSets = function(n, k) {
    const MOD = 1000000007;
    let dp = new Array(n).fill(1);

    for (let iter = 0; iter < k; iter++) {
        const newDp = new Array(n).fill(0);
        let prefix = 0;
        for (let i = 0; i < n; i++) {
            newDp[i] = ((i > 0 ? newDp[i - 1] : 0) + prefix) % MOD;
            prefix = (prefix + dp[i]) % MOD;
        }
        dp = newDp;
    }

    return dp[n - 1];
};